File: vision/detector.py
"""Object detection + tracking on video frames.

Stack: ultralytics (YOLO11) for detection + supervision (ByteTrack) for tracking.
Roboflow `sports` / Roboflow Universe `football-players-detection` provide a trained
model (players/ball/referee). Without a model path we use the default `yolo11n.pt` —
it detects the COCO "sports ball" and "person" classes, which is enough for an
initial pipeline, but production needs a football-specific model.

Heavy imports (torch/ultralytics) are lazy so the rest of the API can start without them.
"""
from collections.abc import Iterator

from vision.types import (
    CLASS_BALL,
    CLASS_GOAL,
    CLASS_PLAYER,
    CLASS_REFEREE,
    Detection,
    FrameResult,
)

# Maps model class names onto our categories. Covers both football-specific models
# (player/ball/referee/goalkeeper) and the default COCO (person/sports ball).
_CLASS_ALIASES = {
    "ball": CLASS_BALL,
    "sports ball": CLASS_BALL,
    "player": CLASS_PLAYER,
    "person": CLASS_PLAYER,
    "referee": CLASS_REFEREE,
    "goalkeeper": "goalkeeper",
    # Stage 2 hook: active only when a football-specific model returns a goal class.
    "goal": CLASS_GOAL,
    "goalpost": CLASS_GOAL,
}


class Detector:
    def __init__(self, model_path: str | None, frame_stride: int = 5, imgsz: int | None = None):
        self.model_path = model_path or "yolo11n.pt"
        self.frame_stride = max(1, frame_stride)
        # Inference resolution. Smaller = faster (big lever on modest GPUs like a
        # GTX 1660). None lets ultralytics pick the model default (usually 640).
        self.imgsz = imgsz
        self._model = None
        self._tracker = None
        self._device = "cpu"

    def _ensure_loaded(self) -> None:
        if self._model is not None:
            return
        from ultralytics import YOLO  # lazy import (heavy)
        import supervision as sv
        import torch

        # Use CUDA when a GPU is present — on CPU, YOLO sustains only a few fps,
        # which is far too slow for live analysis.
        self._device = "cuda" if torch.cuda.is_available() else "cpu"
        self._model = YOLO(self.model_path)
        self._model.to(self._device)
        self._tracker = sv.ByteTrack()
        self._sv = sv

    def _map_class(self, raw_name: str) -> str | None:
        return _CLASS_ALIASES.get(raw_name.lower())

    def detect_frame(self, frame, frame_index: int, timestamp_seconds: float) -> FrameResult:
        """Run detection + tracking on a single BGR numpy frame."""
        import numpy as np
        self._ensure_loaded()
        sv = self._sv

        kwargs = {"verbose": False, "device": self._device}
        if self.imgsz:
            kwargs["imgsz"] = self.imgsz
        result = self._model(frame, **kwargs)[0]
        sv_det = sv.Detections.from_ultralytics(result)
        sv_det = self._tracker.update_with_detections(sv_det)

        detections: list[Detection] = []
        names = result.names
        for j in range(len(sv_det)):
            raw_name = names.get(int(sv_det.class_id[j]), "")
            cls = self._map_class(raw_name)
            if cls is None:
                continue
            x1, y1, x2, y2 = (float(v) for v in sv_det.xyxy[j])
            track_id = (
                int(sv_det.tracker_id[j])
                if sv_det.tracker_id is not None
                else None
            )
            detections.append(
                Detection(
                    cls=cls,
                    xyxy=(x1, y1, x2, y2),
                    confidence=float(sv_det.confidence[j]),
                    track_id=track_id,
                )
            )

        return FrameResult(
            frame_index=frame_index,
            timestamp_seconds=timestamp_seconds,
            detections=detections,
        )

    def run(self, video_path: str) -> Iterator[FrameResult]:
        """Iterates over frames (every `frame_stride`) and yields detections with track_id."""
        self._ensure_loaded()
        sv = self._sv

        frames = sv.get_video_frames_generator(video_path, stride=self.frame_stride)
        info = sv.VideoInfo.from_video_path(video_path)
        fps = info.fps or 25.0

        for i, frame in enumerate(frames):
            frame_index = i * self.frame_stride
            timestamp = frame_index / fps

            kwargs = {"verbose": False, "device": self._device}
            if self.imgsz:
                kwargs["imgsz"] = self.imgsz
            result = self._model(frame, **kwargs)[0]
            sv_det = sv.Detections.from_ultralytics(result)
            sv_det = self._tracker.update_with_detections(sv_det)

            detections: list[Detection] = []
            names = result.names  # {class_id: name}
            for j in range(len(sv_det)):
                raw_name = names.get(int(sv_det.class_id[j]), "")
                cls = self._map_class(raw_name)
                if cls is None:
                    continue
                x1, y1, x2, y2 = (float(v) for v in sv_det.xyxy[j])
                track_id = (
                    int(sv_det.tracker_id[j])
                    if sv_det.tracker_id is not None
                    else None
                )
                detections.append(
                    Detection(
                        cls=cls,
                        xyxy=(x1, y1, x2, y2),
                        confidence=float(sv_det.confidence[j]),
                        track_id=track_id,
                    )
                )

            yield FrameResult(
                frame_index=frame_index,
                timestamp_seconds=timestamp,
                detections=detections,
            )

File: worker-dotnet/PitchWise.Vision/Detector.cs
using OpenCvSharp;
using BT = ByteTrack;

namespace PitchWise.Vision;

/// <summary>
/// Object detection + tracking on video frames. Port of vision/detector.py:
/// YOLO11 (ONNX) for detection + ByteTrack.NET for tracking. Emits
/// <see cref="FrameResult"/> with per-detection track ids and our class names.
///
/// ByteTrack.NET's tracker takes only box+score and returns box+id (it drops the
/// class), so after <c>Update</c> we re-attach each track id to the originating
/// classed detection by best-IoU match — the same information supervision.ByteTrack
/// carries through natively on the Python side.
/// </summary>
public sealed class Detector : IDisposable
{
    /// <summary>Maps a model class name onto our categories (port of _CLASS_ALIASES).
    /// Covers football-specific models and the default COCO (person/sports ball).</summary>
    public static readonly IReadOnlyDictionary<string, string> ClassAliases =
        new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["ball"] = ObjectClass.Ball,
            ["sports ball"] = ObjectClass.Ball,
            ["football"] = ObjectClass.Ball,
            ["soccer ball"] = ObjectClass.Ball,
            ["player"] = ObjectClass.Player,
            ["person"] = ObjectClass.Player,
            ["referee"] = ObjectClass.Referee,
            ["main referee"] = ObjectClass.Referee,
            ["side referee"] = ObjectClass.Referee,
            ["goalkeeper"] = ObjectClass.Goalkeeper,
            ["goal keeper"] = ObjectClass.Goalkeeper,
            // Stage 2 hook: active only when a football model returns a goal class.
            ["goal"] = ObjectClass.Goal,
            ["goalpost"] = ObjectClass.Goal,
        };

    /// <summary>Default class map: alias lookup, drops unknown classes (returns null).</summary>
    public static string? MapClass(string raw) =>
        ClassAliases.TryGetValue(raw, out string? mapped) ? mapped : null;

    private readonly Yolo11OnnxDetector _detector;
    private readonly BT.ByteTracker _tracker;
    private readonly int _frameStride;
    // Re-ID layer on top of ByteTrack's volatile ids. Both null when no Re-ID model is
    // configured (e.g. parity/smoke tests) — the detector then behaves exactly as before.
    private readonly OsNetOnnxEmbedder? _embedder;
    private readonly PlayerReId? _reId;

    /// <param name="modelPath">Exported YOLO11 .onnx.</param>
    /// <param name="classNames">model class-id → raw name (from the exported model / golden).</param>
    /// <param name="frameRate">Source fps; scales the tracker's lost-track buffer.</param>
    /// <param name="frameStride">Process every Nth frame (matches Python frame_stride).</param>
    /// <param name="imgsz">Model input size (must match the export).</param>
    /// <param name="mapClass">Class-name mapper; defaults to <see cref="MapClass"/>.</param>
    /// <param name="reidModelPath">Exported OSNet .onnx for player Re-ID; null disables Re-ID
    /// (detections then carry only volatile TrackId, no stable PlayerId).</param>
    public Detector(
        string modelPath,
        IReadOnlyDictionary<int, string> classNames,
        int frameRate = 25,
        int frameStride = 5,
        int imgsz = 640,
        Func<string, string?>? mapClass = null,
        string executionProvider = "dml",
        int deviceId = 0,
        string? reidModelPath = null)
    {
        _detector = new Yolo11OnnxDetector(
            modelPath, classNames, mapClass ?? MapClass, imgsz,
            executionProvider: executionProvider, deviceId: deviceId);
        var config = new BT.ByteTrackerConfig
        {
            TrackThresh = 0.5f,
            TrackBuffer = 30,
            MatchThresh = 0.8f,
            Mot20 = false,
        };
        // Deterministic track ids across runs (see ByteTrack README).
        BT.BaseTrack.ResetCount();
        _tracker = new BT.ByteTracker(config, frameRate);
        _frameStride = Math.Max(1, frameStride);

        if (!string.IsNullOrWhiteSpace(reidModelPath))
        {
            _embedder = new OsNetOnnxEmbedder(reidModelPath!, executionProvider, deviceId);
            _reId = new PlayerReId();
        }
    }

    /// <summary>Detect + track on a single BGR frame.</summary>
    public FrameResult DetectFrame(Mat frame, int frameIndex, double timestampSeconds) =>
        TrackAndAttach(_detector.Detect(frame), frame, frameIndex, timestampSeconds);

    /// <summary>
    /// Detect on N frames in ONE batched YOLO inference, then run the (stateful, sequential)
    /// tracker frame-by-frame IN ORDER. The batch only groups the expensive YOLO calls; the
    /// tracker sees exactly the same per-frame sequence as N single DetectFrame calls, so
    /// track ids are identical. Frames MUST be passed in ascending frameIndex order.
    /// </summary>
    public IReadOnlyList<FrameResult> DetectFramesBatched(
        IReadOnlyList<(Mat frame, int frameIndex, double timestampSeconds)> batch)
    {
        if (batch.Count == 0) return Array.Empty<FrameResult>();

        var mats = new List<Mat>(batch.Count);
        foreach (var b in batch) mats.Add(b.frame);
        IReadOnlyList<IReadOnlyList<Detection>> perFrame = _detector.DetectBatch(mats);

        var results = new List<FrameResult>(batch.Count);
        for (int k = 0; k < batch.Count; k++)
            results.Add(TrackAndAttach(perFrame[k], batch[k].frame, batch[k].frameIndex, batch[k].timestampSeconds));
        return results;
    }

    /// <summary>Runs the tracker on one frame's detections and re-attaches track ids by IoU,
    /// then (if Re-ID is enabled) assigns a stable PlayerId from the player crops in
    /// <paramref name="frame"/>. Stateful — must be called once per processed frame in
    /// ascending order.</summary>
    private FrameResult TrackAndAttach(IReadOnlyList<Detection> classed, Mat frame, int frameIndex, double timestampSeconds)
    {
        // Feed boxes+scores to the tracker; it returns boxes+ids (no class).
        var btDets = new List<BT.Detection>(classed.Count);
        foreach (Detection d in classed)
            btDets.Add(BT.Detection.FromTlbr(new[] { d.X1, d.Y1, d.X2, d.Y2 }, (float)d.Confidence));

        IReadOnlyList<BT.STrack> tracks = _tracker.Update(btDets);

        // Re-attach track ids to classed detections by best IoU (tracker dropped class).
        var withIds = new List<Detection>(classed.Count);
        var used = new bool[tracks.Count];
        foreach (Detection d in classed)
        {
            double[] box = { d.X1, d.Y1, d.X2, d.Y2 };
            int bestJ = -1;
            double bestIou = 0.5;
            for (int j = 0; j < tracks.Count; j++)
            {
                if (used[j]) continue;
                double iou = Iou(box, tracks[j].Tlbr);
                if (iou > bestIou) { bestIou = iou; bestJ = j; }
            }
            int? id = null;
            if (bestJ >= 0) { used[bestJ] = true; id = tracks[bestJ].TrackId; }
            withIds.Add(d with { TrackId = id });
        }

        IReadOnlyList<Detection> final = AssignPlayerIds(withIds, frame, frameIndex);

        return new FrameResult
        {
            FrameIndex = frameIndex,
            TimestampSeconds = timestampSeconds,
            Detections = final,
        };
    }

    /// <summary>Computes OSNet embeddings for the player crops and maps volatile track ids to
    /// stable PlayerIds via <see cref="PlayerReId"/>. No-op passthrough when Re-ID is disabled.
    /// Embeddings run as ONE batched inference over all player/goalkeeper boxes in the frame.</summary>
    private IReadOnlyList<Detection> AssignPlayerIds(IReadOnlyList<Detection> withIds, Mat frame, int frameIndex)
    {
        if (_embedder is null || _reId is null) return withIds;

        // Embed all detections (index-aligned) so PlayerReId can consume by position; only
        // player-like boxes actually get a crop, others get an empty (no-embedding) vector.
        var toEmbed = new List<Detection>(withIds.Count);
        var embedIndex = new int[withIds.Count];   // maps detection -> slot in toEmbed, or -1
        for (int i = 0; i < withIds.Count; i++)
        {
            Detection d = withIds[i];
            if ((d.Cls == ObjectClass.Player || d.Cls == ObjectClass.Goalkeeper) && d.TrackId is not null)
            {
                embedIndex[i] = toEmbed.Count;
                toEmbed.Add(d);
            }
            else embedIndex[i] = -1;
        }

        IReadOnlyList<float[]> embedded = toEmbed.Count > 0
            ? _embedder.EmbedBatch(frame, toEmbed)
            : Array.Empty<float[]>();

        // Re-expand to a per-detection embedding list (empty vector where no crop was taken).
        var embeddings = new List<float[]>(withIds.Count);
        for (int i = 0; i < withIds.Count; i++)
            embeddings.Add(embedIndex[i] >= 0 ? embedded[embedIndex[i]] : Array.Empty<float>());

        return _reId.Assign(withIds, embeddings, frameIndex);
    }

    /// <summary>Iterates frames (every frame_stride) yielding detections with track ids.
    /// Port of Detector.run(). fps is read from the video for timestamps.</summary>
    /// <param name="startSeconds">Skip decoding up to this timestamp (segment start). 0 = beginning.</param>
    /// <param name="endSeconds">Stop after this timestamp (segment end). null = end of file.</param>
    /// <remarks>Timestamps stay ABSOLUTE relative to the whole video (frameIndex/fps),
    /// so events from a mid-video segment carry their true match time.</remarks>
    public IEnumerable<FrameResult> Run(string videoPath, double startSeconds = 0.0, double? endSeconds = null)
    {
        using var cap = new VideoCapture(videoPath);
        if (!cap.IsOpened())
            throw new InvalidOperationException($"Cannot open video: {videoPath}");

        double fps = cap.Get(VideoCaptureProperties.Fps);
        if (fps <= 0 || double.IsNaN(fps)) fps = 25.0;

        // Seek to the segment start. Snap to a stride boundary so frameIndex%stride==0
        // stays true and track ids/timestamps line up with a from-zero run.
        int startFrame = 0;
        if (startSeconds > 0)
        {
            startFrame = (int)Math.Floor(startSeconds * fps);
            startFrame -= startFrame % _frameStride;
            cap.Set(VideoCaptureProperties.PosFrames, startFrame);
        }
        int? endFrame = endSeconds is double e ? (int)Math.Ceiling(e * fps) : null;

        using var frame = new Mat();
        int srcIndex = startFrame;   // absolute source frame index
        while (cap.Read(frame) && !frame.Empty())
        {
            if (endFrame is int ef && srcIndex >= ef) break;
            if (srcIndex % _frameStride == 0)
            {
                double ts = srcIndex / fps;
                yield return DetectFrame(frame, srcIndex, ts);
            }
            srcIndex++;
        }
    }

    private static double Iou(double[] a, double[] b)
    {
        double ix1 = Math.Max(a[0], b[0]), iy1 = Math.Max(a[1], b[1]);
        double ix2 = Math.Min(a[2], b[2]), iy2 = Math.Min(a[3], b[3]);
        double iw = Math.Max(0.0, ix2 - ix1), ih = Math.Max(0.0, iy2 - iy1);
        double inter = iw * ih;
        double areaA = Math.Max(0.0, a[2] - a[0]) * Math.Max(0.0, a[3] - a[1]);
        double areaB = Math.Max(0.0, b[2] - b[0]) * Math.Max(0.0, b[3] - b[1]);
        double union = areaA + areaB - inter;
        return union <= 0.0 ? 0.0 : inter / union;
    }

    public void Dispose()
    {
        _detector.Dispose();
        _embedder?.Dispose();
    }
}
